src/replicache.ts

```diff
@@ -134,6 +158,167 @@ export default class Replicache implements ReadTransaction {
+      //
+      // TODO: we can rethrow here once replicache-internal is improved to not
+      // treat offlineness as an error.
+      console.info('Error: $e');
+      this._online = false;
+    }
+  }
+
+  protected async _beginSync(): Promise<BeginSyncResult> {
+    const beginSyncResult = await this._invoke('beginSync', {
+      batchPushURL: this._batchUrl,
+      diffServerURL: this._diffServerUrl,
+      dataLayerAuth: this._dataLayerAuth,
+      diffServerAuth: this._diffServerAuth,
+    });
+
+    const {syncInfo} = beginSyncResult;
+
+    // let reauth = false;
+
+    function checkStatus(
+      data: {httpStatusCode?: number; errorMessage?: string},
+      serverName: string,
+    ) {
+      const {httpStatusCode, errorMessage} = data;
+      if (errorMessage !== '') {
+        console.error(
+          `Got error response from ${serverName} server: ${httpStatusCode}: ${errorMessage}`,
+        );
+      }
+      if (httpStatusCode === httpStatusUnauthorized) {
+        // reauth = true;
+      }
+    }
+
+    const {batchPushInfo} = syncInfo;
+    if (batchPushInfo) {
+      checkStatus(batchPushInfo, 'batch');
+      const mutationInfos = batchPushInfo.batchPushResponse.mutationInfos;
+      if (mutationInfos != null) {
+        for (const mutationInfo of mutationInfos) {
+          console.error(
+            `MutationInfo: ID: ${mutationInfo.id}, Error: ${mutationInfo.error}`,
+          );
+        }
+      }
+    }
+
+    checkStatus(syncInfo.clientViewInfo, 'client view');
+
+    // if (reauth && getDataLayerAuth != null) {
+    //   this._dataLayerAuth = await this.getDataLayerAuth();
+    //   if (this._dataLayerAuth != null) {
+    //     // Try again now instead of waiting for another 5 seconds.
+    //     return await this._beginSync();
+    //   }
+    // }
+
+    const syncHead = beginSyncResult.syncHead;
+    const {syncID} = syncInfo;
+    return {syncID, syncHead};
+  }
+
+  protected async _maybeEndSync(
+    beginSyncResult: BeginSyncResult,
+  ): Promise<void> {
+    if (this._closed) {
+      return;
+    }
+
+    let {syncHead} = beginSyncResult;
+
+    const {replayMutations} = await this._invoke(
+      'maybeEndSync',
+      beginSyncResult,
+    );
+    if (!replayMutations || replayMutations.length === 0) {
+      // All done.
+      await this._checkChange(syncHead);
+      return;
+    }
+
+    // Replay.
+    for (const mutation of replayMutations) {
+      const {original} = mutation;
+      syncHead = await this._replay(
+        syncHead,
+        original,
+        mutation.name,
+        mutation.args,
+      );
+    }
+
+    const {syncID} = beginSyncResult;
+    await this._maybeEndSync({syncID, syncHead});
+  }
+
+  private async _replay<A extends JsonType>(
+    basis: string,
+    original: string,
+    name: string,
+    args: A,
+  ): Promise<string> {
+    const mutatorImpl = this._mutatorRegistry.get(name);
+    if (!mutatorImpl) {
```

Comment:
We cannot throw here (does the Flutter code do this?).

Sync must move forward in this case -- log the error and continue. Otherwise a customer can ship a client that omits a mutator accidentally and wedge their customers.

also we must catch, log, and continue any errors that come. up out of mutators during replay.

Reply:
I added this to make TS happy. Let me look at it to see what we actually want to happen here.